staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,417 @@
+		if ok {
+			t.Error("unexpected bool value indicating buffer is not empty")
+		}
+		if event != nil {
+			t.Errorf("unexpected event received, expected: nil, got %v", *event)
+		}
+	}
+}
+
+func TestFillBuffer(t *testing.T) {
+	min := func(a, b int) int {
+		if a < b {
+			return a
+		}
+		return b
+	}
+
+	cases := []struct {
+		numEventsToFill int
+	}{
+		{numEventsToFill: 0},              // no events to put in buffer
+		{numEventsToFill: 5},              // some events to put in buffer < bufferSize
+		{numEventsToFill: bufferSize},     // some events to put in buffer = bufferSize
+		{numEventsToFill: bufferSize + 5}, // some events to put in buffer > bufferSize
+	}
+	for _, c := range cases {
+		events := generateEvents(0, c.numEventsToFill)
+		wci := intervalFromEvents(events)
+
+		eventsProcessed, eventsToProcess := 0, len(events)
+		for eventsToProcess > 0 {
+			wci.fillBuffer()
+			// In one cycle of fillBuffer() you can only receive
+			// bufferSize number of events at the maximum.
+			eventsToRecv := min(bufferSize, eventsToProcess)
+			received := 0
+			for {
+				event, ok := wci.buffer.next()
+				if !ok {
+					if received != eventsToRecv {
+						t.Errorf("unexpected bool value indicating empty buffer %d", c.numEventsToFill)
+					}
+					break
+				}
+				if received == eventsToRecv {
+					t.Errorf("unexpected bool value indicating non-empty buffer")
+					break
+				}
+				if event == nil {
+					t.Errorf("unexpected nil event, expected: %v", *events[received+eventsProcessed])
+					break
+				}
+				if !reflect.DeepEqual(event, events[received+eventsProcessed]) {
+					t.Errorf("unexpected event, expected: %v, got %v", *events[received+eventsProcessed], *event)
+				}
+				received++
+			}
+			eventsToProcess -= received
+			eventsProcessed += received
+		}
+		// Buffer should be empty now, an additional fillBuffer()
+		// should make no difference.
+		wci.fillBuffer()
+		event, ok := wci.buffer.next()
+		if ok {
+			t.Errorf("unexpected boolean value indicating buffer is not empty")
+		}
+		if event != nil {
+			t.Errorf("unexpected event, expected a nil event, got %v", *event)
+		}
+	}
+}
+
+func TestCacheIntervalNextFromWatchCache(t *testing.T) {
+	// Have the capacity such that it facilitates
+	// filling the interval buffer more than once
+	// completely and then some more - 10 here is
+	// arbitrary.
+	const capacity = 2*bufferSize + 10
+	wc := newTestWatchCache(capacity, &cache.Indexers{})
+	for i := 0; i < capacity; i++ {
+		wc.Add(
+			makeTestPod(
+				fmt.Sprintf("pod%d", i+1),
+				uint64(i+1),
+			),
+		)
+	}
+
+	cases := []struct {
+		startIndexWatchCache int
+		endIndexWatchCache   int
+		// offset signifies the offset in the watchCache
+		// starting from which events will be served. The
+		// startIndex of the watchCacheInterval is computed
+		// as startIndexWatchCache + offset which means when
+		// the offset is not specified or is 0, the interval
+		// spans the entire watchCache.
+		offset int
```

Comment:
This isn't exactly what I was asking for.

I would really like "watchCache propagation/creation" to actually be part of test.
And basically propagate watch cache in couple cases where:
- watchcache is partially propagated (with less than capacity items)
- watchcache is fully propagated (with exactly capacity items propagated)
- watchcache is over-propagated (wich more than capacity items propagated into it)

So basically we need the lines 233-239 be part of individual testcase and propagated with test-defined number of items to put into it (parameter of test).

Reply:
So basically, the params I expect here are:
```
eventsAddedToWatchcache int

intervalStartIndex

// This one probably isn't needed at all - it will always be watchcach.EndIndex
intervalEndIndex
```